openedx/xblock-lti-consumer
# feat: LTI 1.3 reusable configuration

## Description

This PR adds compatibility for LTI 1.3 external configurations to the LTI consumer XBlock. This change allow the XBlock to use values from the external LTI configuration when enabled and setup. This PR also includes a fix that adds validation to the external_config field on the LTI consumer XBlock and also improves the behavior of the JS that handles the visibility of configuration fields on the LTI consumer XBlock.

## Type of Change

- [x] Modify `urlpatterns` variable on `plugins/urls.py` module.
- [x] Modify `get_lti_1p3_launch_info` function on `api.py` module.
- [x] Modify `LtiConsumerXBlock` `validate_field_data` method.
- [x] Modify `LtiConsumerXBlock` `_get_lti_block_launch_handler` method. 
- [x] Modify `LtiConsumerXBlock` `_get_lti_1p3_launch_url` method. 
- [x] Modify `LtiConfiguration` `clean` method.
- [x] Modify `LtiConfiguration` `get_lti_advantage_ags_mode` method.
- [x] Modify `LtiConfiguration` `get_lti_advantage_deep_linking_enabled` method.
- [x] Modify `LtiConfiguration` `get_lti_advantage_deep_linking_launch_url` method.
- [x] Modify `LtiConfiguration` `get_lti_advantage_nrps_enabled` method.
- [x] Modify `LtiConfiguration` `get_lti_1p3_redirect_uris` method.
- [x] Modify `LtiConfiguration` `_get_lti_1p3_consumer` method.
- [x] Modify `xblock_studio_view.js`.
- [x] Modify `public_keyset_endpoint` function on `plugin/views.py` module.
- [x] Modify `access_token_endpoint` function on `plugin/views.py` module.
- [x] Include tests for all added and modified code.

## Testing

1. Install and configure openedx-ltistore (follow the setup steps from: https://github.com/openedx/xblock-lti-consumer/pull/239 and use the related PR: https://github.com/open-craft/openedx-ltistore/pull/2)
2. Create an IMS reference tool: https://lti-ri.imsglobal.org/lti/tools
3. Go to http://localhost:18000/admin/lti_store/externallticonfiguration/ and create a new external LTI 1.3 configuration.
4. Fill the "LTI 1.3 Private Key",  "LTI 1.3 Client ID" and "LTI 1.3 Tool Keyset URL" or "LTI 1.3 Tool Public Key" field.
5. Create a course and add a LTI 1.3 consumer XBlock.
6. Set the consumer XBlock to user an external configuration and set the reusable configuration ID (follow the steps on how to get the filter key from: https://github.com/openedx/xblock-lti-consumer/pull/239) 
7. Copy the keyset and access token URL from the XBlock preview back to the IMS reference tool previously created. 
8. Go to the course a launch the XBlock.
9. The LTI 1.3 launch should be valid.
10. Check that the XBlock doesn't allow you to set the external configuration type without adding an external configuration ID.
11. Check that the XBlock doesn't allow you to set an invalid external configuration ID format.
